### PR TITLE
Add memory peak measurements

### PR DESCRIPTION
**Pull request by xadupre**

### CODE REVIEW

File: onnxscript/tools/benchmark/export_model.py [hunk 1]
```diff
@@ -85,6 +86,7 @@ def main(args=None):
     msg = [tuple(i.shape for i in inp) for inp in example_inputs]
     print(f"[export_model] input_shapes={msg}")
     conversion: dict[str, Any] = {}
+    memory_stat: dict[str, float] = {}
```
Comment (justinchuby) [inline onnxscript/tools/benchmark/export_model.py:89]:
```suggestion
    memory_stats: dict[str, float] = {}
```

File: onnxscript/tools/memory_peak.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+import multiprocessing
+import os
+
+
+def get_memory_rss(pid: int) -> int:
+    """
+    Returns the physical memory used by a process.
+
+    :param pid: process id, current one is `os.getpid()`
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:16]: nit: Google stype docstrings

File: onnxscript/tools/memory_peak.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    It relies on the module :epkg:`psutil`.
+    """
+    import psutil
+
+    process = psutil.Process(pid)
+    mem = process.memory_info().rss
+    return mem
+
+
+class Monitor:
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:28]: nit: Looks like it can be a dataclass

File: onnxscript/tools/memory_peak.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+        from onnxscript.tools.memory_peak import start_spying_on, flatten
+
+        p = start_spying_on()
+        # ...
+        # code to measure
+        # ...
+        stat = p.stop()
+        print(stat)
+        print(flatten(stat))
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:227]: nit: Example should come before the Args section (https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: onnxscript/tools/memory_peak.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    """
+    if pid is None:
+        pid = os.getpid()
+    return MemorySpy(pid, delay, cuda)
+
+
+def flatten(ps: dict[str, list[Monitor]], prefix: str = "") -> dict[str, float]:
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:234]: Can this be private?

File: onnxscript/tools/memory_peak.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:4]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
```

File: onnxscript/tools/memory_peak_test.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import os
+import time
+import unittest
+
+import numpy as np
+import torch
+
+from onnxscript.tools.memory_peak import get_memory_rss, start_spying_on
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak_test.py:12]: Import modules only

File: onnxscript/tools/memory_peak_test.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak_test.py:4]: ditto for consistency.

File: onnxscript/tools/memory_peak_test.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import time
+import unittest
+
+import numpy as np
+import torch
+
+from onnxscript.tools.memory_peak import get_memory_rss, start_spying_on
+
+
+class TestMemoryPeak(unittest.TestCase):
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak_test.py:15]: For some reason this test was not run in the CI it seems (judging from the lack of coverage). We need to understand what was going in.

Comment (xadupre) [reply onnxscript/tools/memory_peak_test.py:15]: It is expected. The code is starting another process to monitor the memory. It cannot be done from the same process. The code coverage is not capture in the new process.

File: onnxscript/tools/memory_peak.py (hunk 2 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak.py:14]: done

File: onnxscript/tools/memory_peak.py (hunk 4 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak.py:225]: done

File: onnxscript/tools/memory_peak_test.py (hunk 7 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak_test.py:12]: done

File: onnxscript/tools/memory_peak_test.py (hunk 8 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak_test.py:4]: done

File: onnxscript/tools/memory_peak.py (hunk 5 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak.py:236]: It is used outside this module.

File: onnxscript/tools/benchmark/export_model.py [hunk 10]
```diff
@@ -59,6 +60,7 @@ def main(args=None):
 
     # Import is delayed so that help is being display faster (without having to import heavy packages).
     import onnxscript.tools
+    import onnxscript.tools.memory_peak as mpk
```
Comment (justinchuby) [inline onnxscript/tools/benchmark/export_model.py:63]:
```suggestion
    import onnxscript.tools.memory_peak
```

doesn’t look like it’s used a lot, so

File: onnxscript/tools/memory_peak.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/tools/memory_peak.py:4]: Looks like some changes applied were not reverted? Not sure if my view was outdated. I would double check

File: onnxscript/tools/memory_peak_test.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import os
+import time
+import unittest
+
+import numpy as np
+import torch
+
+import onnxscript.tools.memory_peak as mpeak
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak_test.py:10]:
```suggestion
import onnxscript.tools.memory_peak
```

consistency

**Review by justinchuby — APPROVED** [review]
Final round of minor comments. Thanks!

I noticed some changes marked as done were potentially reverted so I would double check those

File: onnxscript/tools/memory_peak.py (hunk 6 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak.py:4]: Sometimes, it is added by ruff format or isort,  I don't remember. I don't use lintrunner as it does not work for me.

File: onnxscript/tools/benchmark/export_model.py (hunk 10 above)
Comment (xadupre) [reply onnxscript/tools/benchmark/export_model.py]: done

File: onnxscript/tools/memory_peak_test.py (hunk 11 above)
Comment (xadupre) [reply onnxscript/tools/memory_peak_test.py]: done

File: onnxscript/tools/memory_peak.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+import multiprocessing
+import os
+
+
+def get_memory_rss(pid: int) -> int:
+    """
+    Returns the physical memory used by a process.
+
+    :param pid: process id, current one is `os.getpid()`
+    :return: physical memory
```
Comment (justinchuby) [inline onnxscript/tools/memory_peak.py:15]:
```suggestion
    Args:
        pid: Process id, current one is `os.getpid()`.

    Returns:
         Physical memory.
```

**Review by justinchuby — APPROVED** [review]

File: onnxscript/tools/memory_peak.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/tools/memory_peak.py]: That shouldn't have anything to do with the linters. It's just the copyright header

**Pull request merged**